### DISCUSSION TITLE
Custom standard name in output

### DISCUSSION THREAD

**Discussion by kczx3** [PHPCSStandards/PHP_CodeSniffer#290] [Q&A] [answered]
I was working on a standard with some custiom sniffs and I (finally) have it loading and running, however when using an option that outputs the sniff name such as --report-source it isn't outputting the name like I would like.

<details>
<summary>.phpcs.xml</summary>

```xml
<?xml version="1.0"?>
<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="mLinks Coding Standard" xsi:noNamespaceSchemaLocation="vendor/squizlabs/php_codesniffer/phpcs.xsd">
    <description>The coding standard for mLinks.</description>

    <arg name="extensions" value="php" />
    <arg name="basepath" value="."/>

    <!-- Makes it so that warnings don't impact the exit code of phpcs -->
    <config name="ignore_warnings_on_exit" value="1"/>

    <rule ref="./plugins/cs/Mlinks" />

</ruleset>
```
</details>

<details>
<summary>plugins/cs/Mlinks/ruleset.xml</summary>

```xml
<?xml version="1.0"?>
<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="mLinks Coding Standard" namespace="mlinks\cs" xsi:noNamespaceSchemaLocation="../../../vendor/squizlabs/php_codesniffer/phpcs.xsd">
    <description>The coding standard for mLinks.</description>

    <arg name="extensions" value="php" />
    <arg name="basepath" value="."/>

    <rule ref="mlinks.NameValidation.ValidVariableName" />

</ruleset>
```
</details>

When I run this on a file that reports an error and use --report-source, I get the following output:
```
PHP CODE SNIFFER VIOLATION SOURCE SUMMARY
----------------------------------------------------------------------
SOURCE                                                           COUNT
----------------------------------------------------------------------
standard.NameValidation.ValidVariableName.NotCamelCaps           1
----------------------------------------------------------------------
A TOTAL OF 1 SNIFF VIOLATION WERE FOUND IN 1 SOURCE
----------------------------------------------------------------------

Time: 62ms; Memory: 4MB
```

I was hoping that the first part that reads "standard" would instead show the standard's name: Mlinks.  Did I miss this in the documentation as well?

**Comment by jrfnl** [comment]
@kczx3 Please have a look at https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Coding-Standard-Tutorial and https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Annotated-Ruleset (bit at the very top)

The naming conventions have to be strictly followed for things to work as expected.

Based on the above, I can already see two things wrong in the `plugins/cs/Mlinks/ruleset.xml` file:
1. `name="mLinks Coding Standard"`: the name should match the directory name for the standard exactly (and can't have spaces).
2. Standards automatically include all sniffs in the standard, so no need to use `<rule ref="mlinks.NameValidation.ValidVariableName" />`.

Other things to double-check:
1. `namespace="mlinks\cs"`: I can't determine if this is set correctly as your issue contains too little information.
2. `<arg name="basepath" value="."/>`: this is a directive which I would typically recommend to be in the custom ruleset _using_ the standard, not in the standard itself.

If you would have posted a link to the project, I could have given you more specific information. As it is, these are the hints I can give you based on the above.

**Reply by kczx3** [reply]
Apologies but the repository is private with my organization.  So I can just share bits of the code.

I didn't realize the name attribute had to match the directory name of the standard exactly.  Didn't see those requirements listed anywhere.  I have changed the name attribute to be just "Mlinks" however it still shows "standard.NameValidation.ValidVariableName.NotCamelCaps".

What else can I provide to help?

**Reply by kczx3** [reply]
It appears I have accidentally found the resolution.  The namespace that I had my Sniff defined within was `mlinks\cs\standard\Sniffs\NameValidation` which apparently isn't correct.  I removed the `standard\Sniffs\` part and now the Sniff is reporting as I expected/wanted - `mlinks.NameValidation.ValidVariableName.NotSnakeCase`.  This is a little different from the example [DisallowHashCommentsSniff](https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Coding-Standard-Tutorial#disallowhashcommentssniffphp) which does have `Sniffs` within the namespace.

**Reply by jrfnl** [reply, marked as answer]
Hmm... that doesn't sound right.

As per the links I gave you before, there are strict naming requirements, so basically your file, directory and namespace structure should look like this:

Directory | Namespace | Contains | Remark
--- | --- | --- | ---
path/to/mlinks | | `ruleset.xml` | Yes the actual directory name for the standard MUST match the name of the standard
path/to/mlinks/Sniffs/*/ | `mlinks\Sniffs\*` | `*Sniff.php` | The * in directory path + namespace is your category name, like `NameValidation`, the * in the file name is the sniff name, like `ValidVariableName`

If you want a prefix for your namespace name, so `Something\mlinks\Sniffs\Category`, you can indicate that in the `namespace` attribute in the `ruleset.xml` file. For this example, that should be set to `namespace=Something\mlinks`.

Other than that, I can't help you without access to the repo. All of this is mentioned in the docs I linked to (though maybe a bit implicit), so please read those and apply their instructions to the letter.

**Reply by jrfnl** [reply]
Oh and make sure that the `composer.json` file does **not** contain an `autoload` directive for the standard (as it interfers with PHPCS recognizing sniffs by their `Stnd.Category.SniffName` name!

**Reply by kczx3** [reply]
We do have a catch-all in our `composer.json` so perhaps that's the problem.

```json
   "autoload": {
        "psr-4": {
            "": "plugins/"
        }
    }
```

**Comment by kczx3** [comment]
```
plugins
├── cs
│   └── Mlinks
│       ├── ruleset.xml
│       └── Sniffs
│           └── NameValidation
│               ├── ValidFunctionNameSniff.php
│               └── ValidVariableNameSniff.php
```

Namespace for ValidFunctionNameSniff is `cs\mlinks\Sniffs\NameValidation` and the namespace in the ruleset.xml is `cs\mlinks`.

I think I've got it now.

**Reply by jrfnl** [reply]
Glad it helped. If you have suggestions on how the docs can be improved, please let me know.
